src/celeritas/phys/RootPrimaryGenerator.cc

```diff
@@ -0,0 +1,121 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/phys/RootPrimaryGenerator.cc
+//---------------------------------------------------------------------------//
+#include "RootPrimaryGenerator.hh"
+
+#include <TFile.h>
+#include <TLeaf.h>
+#include <TTree.h>
+
+#include "corecel/Macros.hh"
+#include "corecel/cont/Range.hh"
+#include "corecel/io/Logger.hh"
+#include "celeritas/ext/ScopedRootErrorHandler.hh"
+#include "celeritas/phys/ParticleParams.hh"
+#include "celeritas/phys/Primary.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from input file, number of events, and primaries to be sampled per
+ * event.
+ */
+RootPrimaryGenerator::RootPrimaryGenerator(std::string const& filename,
+                                           SPConstParticles params,
+                                           size_type num_events,
+                                           size_type primaries_per_event,
+                                           unsigned int seed)
+    : params_(std::move(params))
+    , num_events_{num_events}
+    , primaries_per_event_(primaries_per_event)
+{
+    CELER_EXPECT(!filename.empty());
+    CELER_VALIDATE(num_events > 0 && primaries_per_event > 0,
+                   << "'num_events' and 'primaries_per_event options in input "
+                      "file must be set and positive");
```

Comment:
Validate message should first say what's wrong, not what it should be.
```suggestion
    CELER_VALIDATE(num_events > 0, << "invalid 'num_events' (should be positive)");
    CELER_VALIDATE(primaries_per_event > 0, << "invalid 'primaries_per_event'  (should be positive)");
```